File: python/problem41.py
import math

def largestNLengthPandigitalPrime(n):

	startIndex = int(math.pow(10, n-1)) + 1
	endIndex = int(math.pow(10,n)) - 1

	maxNLengthPandigitNum = 1
	for i in range(startIndex, endIndex):

		if isPandigital(i, n) and isPrime(i):
			if i > maxNLengthPandigitNum:
				maxNLengthPandigitNum = i

	return maxNLengthPandigitNum


def isPandigital(num, uptoN):
    numAsStr = str(num)
    digits = ["1", "2", "3", "4", "5", "6", "7", "8", "9"]
    for i in range(0, uptoN):
        if numAsStr.find(digits[i]) == -1:
            return False

    return True


def isPrime(num):
    if num <= 1:
        return False

    if num == 2:
        return True

    if num % 2 == 0:
        return False

    for i in range(3, int(math.sqrt(num)) + 1, 2):
        if num % i == 0:
            return False

    return True


if __name__=="__main__":
    print(largestNLengthPandigitalPrime(4))

File: go/problem41.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(LargestNLengthPandigitalPrime(4))
}

func LargestNLengthPandigitalPrime(n int) int {

	var startStr strings.Builder
	for i := 1; i <= n; i++ {
		temp := strconv.Itoa(i)
		startStr.WriteString(temp)
	}
	startIndex, err := strconv.Atoi(startStr.String())
	if err != nil {
		fmt.Println("Error in start string to int conversion ", err)
		return -1
	}
	
	endIndex := 0
	orig_number := startIndex
	for orig_number > 0 {
		endIndex = endIndex*10 + orig_number%10
		orig_number /= 10
	}

	maxNLengthPandigitNum := 1
	for i := startIndex; i < endIndex; i++ {

		if isPandigital(i, n) && isPrime(i) {
			if i > maxNLengthPandigitNum {
				maxNLengthPandigitNum = i
			}
		}
	}
	return maxNLengthPandigitNum
}

func isPandigital(num int, uptoN int) bool {

	numAsStr := strconv.Itoa(num)

	digits := []string{"1", "2", "3", "4", "5", "6", "7", "8", "9"}
	for i := 0; i < uptoN; i++ {
		if strings.Index(numAsStr, digits[i]) == -1 {
			return false
		}
	}
	return true
}

func isPrime(num int) bool {

	if num <= 1 {
		return false
	}

	if num == 2 {
		return true
	}

	if num%2 == 0 {
		return false
	}

	for i := 3; i <= int(math.Sqrt(float64(num))); i += 2 {
		if num%i == 0 {
			return false
		}
	}

	return true
}
